File: tests/manual/dev_smoke.c
/* End-to-end /dev exercise: a real C program using the v4 character devices
 * through the full libc + wasm syscall path. Driven by tests/manual/dev_smoke.js.
 *
 * Lives in tests/manual/ rather than tests/unit/ because /dev only exists on a
 * v4 BLOCK_FS mount (the C-level unit harness also runs on v3, and run.py uses
 * the host Node fs where "/dev/*" would hit the real OS) — so it needs a v4
 * mount set up explicitly, which the harness does.
 */
#include <stdio.h>
#include <fcntl.h>
#include <unistd.h>
#include <string.h>
#include <sys/stat.h>
#include <sys/sysmacros.h>

int main(void) {
  struct stat st;

  /* /dev/zero is a character device with the expected device number. */
  if (stat("/dev/zero", &st) != 0) { printf("FAIL stat /dev/zero\n"); return 1; }
  printf("zero_ischr %d\n", S_ISCHR(st.st_mode) ? 1 : 0);
  printf("zero_dev %d:%d\n", major(st.st_rdev), minor(st.st_rdev));

  /* Read from /dev/zero → all zero bytes. */
  int z = open("/dev/zero", O_RDONLY);
  if (z < 0) { printf("FAIL open /dev/zero\n"); return 2; }
  unsigned char b[64];
  memset(b, 0xAB, sizeof b);
  ssize_t n = read(z, b, sizeof b);
  int all_zero = (n == (ssize_t)sizeof b);
  for (size_t i = 0; i < sizeof b; i++) if (b[i] != 0) all_zero = 0;
  printf("zero_read_ok %d\n", all_zero);
  close(z);

  /* Write to /dev/null → swallowed, returns the full count. */
  int nul = open("/dev/null", O_WRONLY);
  if (nul < 0) { printf("FAIL open /dev/null\n"); return 3; }
  ssize_t w = write(nul, "discard me", 10);
  printf("null_write %d\n", w == 10 ? 1 : 0);
  /* And /dev/null reads as immediate EOF. */
  ssize_t re = read(open("/dev/null", O_RDONLY), b, sizeof b);
  printf("null_eof %d\n", re == 0 ? 1 : 0);
  close(nul);

  /* /dev/urandom yields varied bytes. */
  int r = open("/dev/urandom", O_RDONLY);
  if (r < 0) { printf("FAIL open /dev/urandom\n"); return 4; }
  unsigned char rb[256];
  memset(rb, 0, sizeof rb);
  read(r, rb, sizeof rb);
  int nonzero = 0;
  for (size_t i = 0; i < sizeof rb; i++) if (rb[i] != 0) nonzero++;
  printf("urandom_varied %d\n", nonzero > 0 ? 1 : 0);
  close(r);

  printf("DONE\n");
  return 0;
}

File: tests/manual/dev_smoke.js
#!/usr/bin/env node
// End-to-end /dev smoke test: compiles tests/manual/dev_smoke.c with the real
// compiler and runs it against a fresh v4 BLOCK_FS (which auto-creates /dev),
// asserting the program's self-reported results. Lives in tests/manual/ (not
// the unit suite) because /dev is v4-only — see the note in dev_smoke.c.
//
//   node tests/manual/dev_smoke.js

'use strict';
const fs = require('fs');
const path = require('path');
const ROOT = path.resolve(__dirname, '..', '..');

const compiler = require(path.join(ROOT, 'compiler.js'));
const runModule = require(path.join(ROOT, 'host.js'));
const BLOCK_FS = runModule.BLOCK_FS;

function fail(m) { console.error(`FAIL: ${m}`); process.exit(1); }

async function main() {
  const cFile = path.join(__dirname, 'dev_smoke.c');
  const pp = compiler.createDefaultPPRegistry();
  pp.fileReader = (p) => { try { return fs.readFileSync(p, 'utf-8'); } catch { return null; } };
  const compilerOptions = {
    debugSwitch: false, allowImplicitInt: false, allowEmptyParams: false,
    allowKnRDefinitions: false, allowImplicitFunctionDecl: false,
    allowUndefined: false, gcSections: false, gcNoExportRoots: false,
    noUndefined: false, requireSources: [], backend: 'default',
  };
  const warningFlags = { pointerDecay: false, circularDependency: false, largeStackFrame: false };
  const errBuf = [];
  const writeErr = (s) => errBuf.push(Buffer.from(s));

  let wasm;
  try {
    const units = compiler.parseAllUnits(fs, pp, [path.relative(ROOT, cFile)], { warningFlags, compilerOptions, writeErr });
    const link = compiler.linkTranslationUnits(units, compilerOptions);
    if (link.errors.length) fail(`link errors:\n${link.errors.map(e => e.message).join('\n')}`);
    wasm = compiler.generateCode(units, 'dev_smoke.wasm', { compilerOptions, warningFlags, writeErr });
  } catch (e) {
    fail(`compile threw: ${e.message}\n${Buffer.concat(errBuf)}`);
  }

  const store = new BLOCK_FS.MemoryByteStore(1 << 20);
  const blockFS = BLOCK_FS.createV4(store);
  const out = [];
  const exitCode = await runModule({
    bytes: wasm, args: ['dev_smoke.wasm'], env: {},
    writeOut: (b) => out.push(Buffer.from(b)),
    writeErr: (b) => out.push(Buffer.from(b)),
    blockFsFactory: async (ctx) => ({ c: blockFS.toWasmEnv(ctx) }),
  });
  const stdout = Buffer.concat(out).toString('utf-8');
  process.stdout.write(stdout.replace(/^/gm, '  | '));

  const want = [
    'zero_ischr 1', 'zero_dev 1:5', 'zero_read_ok 1',
    'null_write 1', 'null_eof 1', 'urandom_varied 1', 'DONE',
  ];
  for (const line of want) if (!stdout.includes(line)) fail(`missing output line: "${line}"`);
  if (exitCode !== 0) fail(`program exited ${exitCode}`);
  console.log('PASS');
}

main().catch((e) => { console.error(e.stack || e.message); process.exit(1); });
